UI/Controls/NodeTree/TreeView.cs

```diff
@@ -0,0 +1,304 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using BhModule.Community.Pathing.State;
+using BhModule.Community.Pathing.UI.Controls.TreeNodes;
+using BhModule.Community.Pathing.Utility;
+using Blish_HUD;
+using Blish_HUD.Content;
+using Blish_HUD.Controls;
+using Microsoft.Xna.Framework;
+using TmfLib.Pathable;
+
+namespace BhModule.Community.Pathing.UI.Controls.TreeView
+{
+    public class TreeView : Container
+    {
+        private Control _scrollToChildControl = null;
+
+        private PathingCategoryNode _rootNode;
+        private static readonly Logger _logger = Logger.GetLogger<TreeView>();
+
+        public  PackInitiator          PackInitiator  { get; private set; }
+        public  IList<TreeNodeBase>    AllBaseNodes   { get; }      = new List<TreeNodeBase>();
+        public  IList<TreeNodeBase>    ChildBaseNodes { get; }      = new List<TreeNodeBase>();
+        private IList<PathingCategory> AllCategories  { get; set; } = new List<PathingCategory>();
+
+        public event EventHandler<EventArgs> NodeLoadingStarted;
+        public event EventHandler<EventArgs> NodesLoadedFinished;
+
+        public TreeView(PackInitiator packInitiator) {
+            PackInitiator                                                                   =  packInitiator;
+        }
+
+        public void AddNode(TreeNodeBase node) {
+            if(!AllBaseNodes.Contains(node))
+                AllBaseNodes.Add(node);
+        }
+
+        public void RemoveNode(TreeNodeBase node) {
+            if(AllBaseNodes.Contains(node))
+                AllBaseNodes.Remove(node);
+        }
+
+        public void SetPackInitiator(PackInitiator packInitiator) {
+            if (packInitiator == null || PackInitiator == packInitiator) return;
+
+            PackInitiator = packInitiator;
+
+            PackInitiator.PackState.UserConfiguration.GlobalPathablesEnabled.SettingChanged += GlobalPathablesEnabledOnSettingChanged;
+        }
+
+        protected override void OnChildAdded(ChildChangedEventArgs e)
+        {
+            if (!(e.ChangedChild is TreeNodeBase newChild)) return;
+
+            AddNode(newChild);
+            this.ChildBaseNodes.Add(newChild);
+
+            ReflowChildLayout(this.ChildBaseNodes);
+
+            base.OnChildAdded(e);
+        }
+
+        protected override void OnChildRemoved(ChildChangedEventArgs e)
+        {
+            if (e.ChangedChild is TreeNodeBase newChild) {
+                RemoveNode(newChild);
+                this.ChildBaseNodes.Remove(newChild);
+            }
+
+            base.OnChildRemoved(e);
+        }
+
+        public override void RecalculateLayout() {
+            try {
+                ReflowChildLayout(ChildBaseNodes);
+            } catch (Exception ex) {
+                //Investigate why collection is sometimes modified during reflow
+                _logger.Info($"Could not recalculate TreeView layout: {ex.Message}");
```

Comment:
Why is this an INFO?

Reply:
I've now changed it to be a warning. 

It's not critical error, and it seems to be quite rare, which is why I haven't fully investigated it yet.